model/flow/ledger.go

```diff
@@ -9,14 +9,65 @@ import (
 	"github.com/onflow/flow-go/model/fingerprint"
 )
 
+var hexValues []byte = make([]byte, 512)
+
+func init() {
+	// create the 'hexValues' byte slice for use with Key
+	allchars := make([]byte, 256)
+	for i := 0; i < 256; i++ {
+		allchars[i] = byte(i)
+	}
+
+	// this is a table of all hex values
+	hex.Encode(hexValues, allchars)
+}
+
 type RegisterID struct {
 	Owner      string
 	Controller string
 	Key        string
 }
 
+// this function returns a string format of a RegisterID in the form '%x/%x/%x'
+// it has been optimized to avoid the memory allocations inside Sprintf
 func (r *RegisterID) String() string {
-	return fmt.Sprintf("%x/%x/%x", r.Owner, r.Controller, r.Key)
+	ownerLen := len(r.Owner)
+	controllerLen := len(r.Controller)
+	keyLen := len(r.Key)
+
+	requiredLen := ((ownerLen + controllerLen + keyLen) * 2) + 2
+
+	arr := make([]byte, requiredLen)
+	n := 0
+
+	// we're converting characters in the string into the exact 2-character representation in hex
+	for i := 0; i < ownerLen; i++ {
+		arr[n] = hexValues[int(r.Owner[i])*2]
```

Comment:
do we know how much faster doing this custom hex encoding is, rather than using hex.EncodeToString? that would still avoid using the fmt package, and seems a little simpler than this.

I see that https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/encoding/hex/hex.go;l=98 is still using a hex table https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/encoding/hex/hex.go;drc=refs%2Ftags%2Fgo1.17.3;l=15 though they encode on char at a time.

Reply:
It looks like the hex.Encode() will be at least as fast as the new code, so I'm going to switch to using that.  The compiler must have inserted boundary-checking code.
It will still be faster than the original by a significant degree.